### ISSUE TITLE
Train transfer weights issue

### ISSUE THREAD

**Issue by louisck27** [ultralytics/ultralytics#4177]
### Search before asking

- [ ] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

How can keep the origin data class?
I want to add a new class [fan] to the existing model [yolov8n.pt] and remain all existing yolov8n.pt] class.
 **when i load my model [best.pt] it just can detect my new class. losted all previous classes **
do i need to add  coco128\images\train2017 image+label to my train directory "C:\fan\Train\images"?
![image](https://github.com/ultralytics/ultralytics/assets/140987347/b6db7039-268d-4614-afaf-7f19c7f9783b)
edit the nc:80 to 81
model = YOLO('yolov8n.yaml').load('yolov8n.pt')  # build from YAML and transfer weights
model.train(data='newmodel.yaml',imgsz=640,epochs=50, batch=5)

C:\fan\Train\images  _# only fan image_
C:\fan\Ver\images  _# only fan image +  coco128\images\train2017 in (coco128.zip)_

In my train result directory,it seems to works
![image](https://github.com/ultralytics/ultralytics/assets/140987347/3230f4e7-acb2-46cf-8f6f-cbce33d4471d)
'newmodel.yaml'
names 0-79 all detected fail , but 80 (new class) is work
![image](https://github.com/ultralytics/ultralytics/assets/140987347/20b6cb1f-e2db-4154-862d-4c5ada00a0ad)

**Comment by louisck27** [comment]
> ### Search before asking
> * [ ]  I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.
> 
> ### Question
> How can keep the origin data class? I want to add a new class [fan] to the existing model [yolov8n.pt] and remain all data. **when i load my model [best.pt] it just can detect my new class. lost all previous classes ** do i need to add coco128\images\train2017 image+label to my train directory "C:\fan\Train\images"? ![image](https://user-images.githubusercontent.com/140987347/258576963-b6db7039-268d-4614-afaf-7f19c7f9783b.png) edit the nc:80 to 81 model = YOLO('yolov8n.yaml').load('yolov8n.pt') # build from YAML and transfer weights model.train(data='newmodel.yaml',imgsz=640,epochs=50, batch=5)
> 
> C:\fan\Train\images _# only fan image_ C:\fan\Ver\images _# only fan image + coco128\images\train2017 in (coco128.zip)_
> 
> In my train result directory,it seems to works ![image](https://user-images.githubusercontent.com/140987347/258577251-3230f4e7-acb2-46cf-8f6f-cbce33d4471d.png) 'newmodel.yaml' names 0-79 all detected fail , but 80 (new class) is work ![image](https://user-images.githubusercontent.com/140987347/258578032-20b6cb1f-e2db-4154-862d-4c5ada00a0ad.png)
> 
> ### Additional
> _No response_
For additional information to my question:
do i need to add coco128\images\train2017 image+label to my train directory "C:\fan\Train\images"?
when i put all image to my train directory [coco128\images\train2017] -> [C:\fan\Train\images]
However after the training,the result is i can see the previous class but the score is lower.. do I need to add more training step epochs=100, batch=10?
model.train(data='newmodel.yaml',imgsz=640,epochs=50, batch=5)
model = YOLO('yolov8n.pt')
![image](https://github.com/ultralytics/ultralytics/assets/140987347/afe7292d-f9f2-4949-8501-0d2df2ae745e)
model = YOLO('best.pt')
![image](https://github.com/ultralytics/ultralytics/assets/140987347/6bc28ad4-c2a3-4c26-be02-945eee3a7289)

**Comment by louisck27** [comment]
I follow the train step below.

from ultralytics import YOLO

# Load a model
model = YOLO('yolov8n.yaml')  # build a new model from YAML
model = YOLO('yolov8n.pt')  # load a pretrained model (recommended for training)
model = YOLO('yolov8n.yaml').load('yolov8n.pt')  # build from YAML and transfer weights

# Train the model
model.train(data='coco128.yaml', epochs=100, imgsz=640)

transfer weights not work i guss.
getting different result
![image](https://github.com/ultralytics/ultralytics/assets/140987347/b0348e31-ee1d-4343-931d-c608a2246d78)

**Comment by glenn-jocher** [comment]
@louisck27 based on your script, you are trying to load the 'yolov8n.pt' pretrained weights into the model and then train it further. This should normally work as you expected, allowing the model to use the learned weights and continue improving on top of them. In your case, since you noticed a difference in performance, it might be due to several reasons:

1. **Dataset differences:** If the new datasets you are training on are significantly different from the original dataset used to pretrain 'yolov8n.pt', the pretrained weights might not be as effective, and the model might need more epochs to adapt to the new data.

2. **Training parameters:** The batch size, learning rate, and other training parameters might be different from the ones used to train 'yolov8n.pt'. These differences might affect the effectiveness of the pretrained weights.

3. **Model differences:** If there are differences between 'yolov8n.yaml' and the model architecture used to pretrain 'yolov8n.pt', the weights might not be transferred appropriately.

Therefore, I would suggest to double-check each of these possibilities. If you still encounter the same issue, consider training the model from scratch without the pretrained weights and see if there's an improvement.

**Comment by louisck27** [comment]
> @louisck27 based on your script, you are trying to load the 'yolov8n.pt' pretrained weights into the model and then train it further. This should normally work as you expected, allowing the model to use the learned weights and continue improving on top of them. In your case, since you noticed a difference in performance, it might be due to several reasons:
> 
> 1. **Dataset differences:** If the new datasets you are training on are significantly different from the original dataset used to pretrain 'yolov8n.pt', the pretrained weights might not be as effective, and the model might need more epochs to adapt to the new data.
> 2. **Training parameters:** The batch size, learning rate, and other training parameters might be different from the ones used to train 'yolov8n.pt'. These differences might affect the effectiveness of the pretrained weights.
> 3. **Model differences:** If there are differences between 'yolov8n.yaml' and the model architecture used to pretrain 'yolov8n.pt', the weights might not be transferred appropriately.
> 
> Therefore, I would suggest to double-check each of these possibilities. If you still encounter the same issue, consider training the model from scratch without the pretrained weights and see if there's an improvement.

But if I just have my new class "fan" images for training can still transfer success?
** does not include all train images in coco128.zip

Besides my failed script only can detect my new class "fan"
model = YOLO('yolov8n.yaml').load('yolov8n.pt') # build from YAML and transfer weights
model.train(data='coco128.yaml', epochs=100, imgsz=640)

C:\fan\Train\images # only fan image
C:\fan\Ver\images # only fan image

**Comment by glenn-jocher** [comment]
@louisck27 adding a new class and retraining the model with only the new class images without including all original training images might affect the performance in recognizing the original classes. Training the model only with the "fan" class images will just fine-tune the model to your specific "fan" class, possibly at the expense of the other classes.

It's important to note here that when you fine-tune with only the new class "fan", the model would majorly focus on learning features for this class and might not suitably predict the other existing classes because it lacks the required recent training data about them.

If you want to keep the original classes and add a new class without losing performance, you should include the original training data along with the new class data, giving the model the opportunity to learn about all the classes together. You might need to balance the new added class images with the original data to prevent the model from overfitting to the "fan" class.

**Comment by louisck27** [comment]
> If you want to keep the original classes and add a new class without losing performance, you should include the original training data along with the new class data, giving the model the opportunity to learn about all the classes together. You might need to balance the new added class images with the original data to prevent the model from overfitting to the "fan" class.

thank you for the reply, i will try to include the original training data along.

**Comment by glenn-jocher** [comment]
@louisck27 you're welcome! Yes, include the original training data along with your new class data should help the model to maintain its performance on the original classes while learning the new "fan" class. Remember to have a balanced dataset so that the model does not excessively learn the features of the dominant class. Feel free to share your results or any further questions you may have. Good luck with your training!
